gdal/frmts/wcs/wcsutils.cpp

```diff
@@ -0,0 +1,698 @@
+    if (data) {
+        for (int i = 0; data[i]; ++i) {
+            char *val = strchr(data[i], '=');
+            if (*val == '=') {
+                *val = '\0';
+                val = val + 1;
+                if ((key != "" && key == data[i])
+                    || (value != "" && value == val)
+                    || (strcmp(data[i], "foo") == 0))
+                {
+                    if (key != "") {
+                        filename = data[i];
+                    } else if (value != "") {
+                        filename = data[i];
+                    }
+                    continue;
+                }
+                data2 = CSLAddNameValue(data2, data[i], val);
+            }
+        }
+        CSLDestroy(data);
+    }
+    CSLSave(data2, db_name); // returns 0 in error and for empty arrays
+    CSLDestroy(data2);
+    if (filename != "") {
+        char **folder = VSIReadDir(cache_dir);
+        int size = folder ? CSLCount(folder) : 0;
+        for (int i = 0; i < size; i++) {
+            if (*folder[i] == '.') {
+                continue;
+            }
+            CPLString name = folder[i];
+            if (name.find(filename) != std::string::npos) {
+                CPLString filepath = CPLFormFilename(cache_dir, name, NULL);
+                remove(filepath);
+            }
+        }
+        CSLDestroy(folder);
+    }
+    return true;
+}
+
+/* -------------------------------------------------------------------- */
+/*      FromCache                                                       */
+/* -------------------------------------------------------------------- */
+
+int FromCache(CPLString cache_dir, CPLString &filename, CPLString url)
+{
+    int retval = 0;
+    CPLString db_name = CPLFormFilename(cache_dir, "db", NULL);
+    VSILFILE *db = VSIFOpenL(db_name, "r");
+    if (db) {
+        while (const char *line = CPLReadLineL(db)) {
+            char *value = strchr((char *)line, '=');
+            if (*value == '=') {
+                *value = '\0';
+            } else {
+                continue;
+            }
+            if (strcmp(url, value + 1) == 0) {
+                filename = line;
+                retval = 1;
+                break;
+            }
+        }
+        VSIFCloseL(db);
+    }
+    if (retval == 0) {
+        db = VSIFOpenL(db_name, "a");
+        if (!db) {
+            CPLError(CE_Failure, CPLE_FileIO, "Can't open file '%s': %i\n", db_name.c_str(), errno);
+            return -1;
+        }
+        // using tempnam is not a good solution
+        char *path = tempnam(cache_dir, "");
+        filename = CPLGetFilename(path);
+        free(path);
+        CPLString entry = filename + "=" + url + "\n"; // '=' for compatibility with CSL
+        VSIFWriteL(entry.c_str(), sizeof(char), entry.size(), db);
+        VSIFCloseL(db);
+    }
+    filename = CPLFormFilename(cache_dir, filename, NULL);
+    return retval;
+}
+
+CPLXMLNode *AddSimpleMetaData(char ***metadata,
```

Comment:
const CPLString& from. And likely const CPLString& path as well, unless the intent is really to return a modified version to the caller ?